### PR TITLE
WIP: Training Data Monitoring

### PR DESCRIPTION
**Pull request by rizar**
In this pull request I implement what we typically call per-batch monitoring.

In order to do that the following changes are done:

* I separate a class `AggregationBuffer` from `DatasetEvaluator`. The new class is in fact simply a collection of aggregators and is agnostic to the way data is iterated.

* The `AggregationBuffer` is used in the `TrainingDataMonitoring` to do the job.

What I find quite cool about the way it is currently done is the user can control the period over which data is averaged by configuring the extension.

I plan to do a bit more before merging:

- [x] Add some more documentation to the extension.

- [x] Test the extension.

### CODE REVIEW

**Comment by rizar** [comment]
Stumbled upon an interesting situation: if I do per-batch monitoring, average results and commit results to the log only at the end of the epoch, what should be my policy with data independent properties? E.g. should I report the average gradient norm during the epoch or the gradient norm after the last iteration? What behaviour should be default?

**Comment by rizar** [comment]
Okay, there first funny issues with monitoring / aggregation

1. Current approach for aggregation schemes does not really work for anything except scalars. The main problem is that the real shape of the monitored tensor is only available with the data. Therefore we need to somehow change the shape of the accumulator when data arrives. After a number of tries and failures I gave it up, though I am pretty sure a solution exists. 

2. Depending on whether an aggregation scheme is used, data independent expressions (such as sum of matrix elements) are read out at different stages of training (see test_monitoring.py for an illustration of the problem). I think I know how to fix this one, by perhaps I would merge first.

What do you guys think about this pull request? 

@bartvm, @janchorowski

**Comment by janchorowski** [comment]
On 1/21/2015 9:20 PM, Dmitry Bogdanov wrote:
>
> Okay, there first funny issues with monitoring / aggregation
>
> 1.
>
>     Current approach for aggregation schemes does not really work for
>     anything except scalars. The main problem is that the real shape
>     of the monitored tensor is only available with the data. Therefore
>     we need to somehow change the shape of the accumulator when data
>     arrives. After a number of tries and failures I gave it up, though
>     I am pretty sure a solution exists.
>
Only the number of dimensions has to match, it is known at compile time. 
The exact shape of a shared variable can change during run. At least it 
works for trainers in GH.

**Comment by rizar** [comment]
That's true, but it is a bit more involved in practice. You can try yourself. I have an idea how to fix, though, but it is not worth telling before trying.

**Comment by rizar** [comment]
Okay, the second weird issue seems to be combated by introducing another aggregation scheme. Now values of all expressions are computed right before another parameters are changed by the training algorithm.

Fixing tensor aggregation is rather out of scope of this pull request, so I will just create an issue.

File: blocks/algorithms/__init__.py [hunk 1]
```diff
@@ -169,6 +170,9 @@ def __init__(self, step_rule=None, gradients=None, **kwargs):
                 zip(self.params, tensor.grad(self.cost, self.params))))
         self.step_rule = step_rule if step_rule else SteepestDescent()
 
+        self.total_gradient_norm = named_copy(
+            sum(g.norm(2) for g in self.gradients), "total_gradient_norm")
```
Comment (bartvm) [inline blocks/algorithms/__init__.py]: Should probably be `tensor.sqrt(sum(g ** 2 for g in self.gradients))`; the L2 norm isn't additive.

File: blocks/monitoring/evaluators.py [hunk 2]
```diff
@@ -4,20 +4,130 @@
...
+from blocks.monitoring.aggregation import _DataIndependent, Mean, TakeLast
 from blocks.graph import ComputationGraph
 
 logger = logging.getLogger()
 
 
+class AggregationBuffer(object):
+    """Intermediate results of aggregating values of Theano expressions.
+
+    Incapsulates aggregators for a list of Theano expressions. Collects
```
Comment (bartvm) [inline blocks/monitoring/evaluators.py]: *encapsulates

File: blocks/algorithms/__init__.py (hunk 1 above)
Comment (rizar) [inline blocks/algorithms/__init__.py]: Good point!

**Comment by rizar** [comment]
The pull request is ready to merge, but I have a weird problem with importing PyLearn2 in one out of 4 test runs. @bartvm , can you look?

**Comment by bartvm** [comment]
`fatal: read error: Connection reset by peer`, looks like Pylearn2 installation failed because of a connection error. I restarted it.

File: blocks/algorithms/__init__.py [hunk 3]
```diff
@@ -169,6 +170,10 @@ def __init__(self, step_rule=None, gradients=None, **kwargs):
                 zip(self.params, tensor.grad(self.cost, self.params))))
         self.step_rule = step_rule if step_rule else SteepestDescent()
 
+        self.total_gradient_norm = named_copy(
```
Comment (bartvm) [inline blocks/algorithms/__init__.py]: Maybe on the long term it would be useful to store all these variables in an attribute `TrainingAlgorithm.monitoring_channels` (or whatever term we want to use). In that case a user can just go through the algorithms `monitoring_channels` attribute to see what the algorithm suggests in terms of monitoring values. That's easier than having to go through the documentation of each algorithm to see what monitors it makes public that you could monitor.

Comment (rizar) [inline blocks/algorithms/__init__.py]: Yep, I also had it mind. I see no problem in adding it a bit later, and seems we can release 0.01 without it.

Comment (bartvm) [inline blocks/algorithms/__init__.py]: Cool, can you make a ticket for it? (I'm on my phone)

**Comment by rizar** [comment]
Any objections to merging this guy?

**Comment by bartvm** [comment]
Nope!
